src/api/mod.rs

```diff
@@ -0,0 +1,119 @@
+/*!
+
+Interface to local and remote PGXN mirrors and the PGXN API.
+
+*/
+use crate::error::BuildError;
+use std::{fs::File, io, path::Path, time::Duration};
+use url::Url;
+
+/// Interface to the PGXN API.
+pub struct Api {
+    url: String,
+    agent: ureq::Agent,
+}
+
+impl Api {
+    /// Creates a new Api to access the PGXN API at `url`. Supports `file:`
+    /// and `https:` URLs. Pass `proxy` to proxy requests. Returns a
+    /// BuildError::Http if the Proxy URL is invalid.
+    pub fn new(url: String, proxy: Option<String>) -> Result<Api, BuildError> {
+        let mut builder = ureq::AgentBuilder::new()
+            .timeout_read(Duration::from_secs(5))
+            .timeout_write(Duration::from_secs(5))
+            .https_only(true)
+            .user_agent("pgxn-http/1.0");
+
+        if let Some(p) = proxy {
+            builder = builder.proxy(ureq::Proxy::new(&p)?);
+        }
+
+        Ok(Api {
+            url,
+            agent: builder.build(),
+        })
+    }
+
+    /// Download version `version` of `dist` to `dir`.
+    pub fn download_to<P: AsRef<Path>>(
+        &self,
+        dir: P,
+        dist: &str,
+        version: &str,
+    ) -> Result<(), BuildError> {
+        // TODO: use URI templates.
+        let url = format!("{}/dist/{dist}/{version}/{dist}-{version}.zip", self.url);
+        let url = Url::parse(&url)?;
+        self.download_url_to(dir, url)
+    }
+
+    /// Download `url` to `dir`. The file name must be the last segment of the URL.
+    fn download_url_to<P: AsRef<Path>>(&self, dir: P, url: url::Url) -> Result<(), BuildError> {
+        // Extract the file name from the URL.
+        match url.path_segments() {
+            None => Err(BuildError::NoUrlFile(url))?,
+            Some(segments) => match segments.last() {
+                None => Err(BuildError::NoUrlFile(url))?,
+                Some(filename) => {
+                    if filename.is_empty() {
+                        return Err(BuildError::NoUrlFile(url));
+                    }
+                    let dst = dir.as_ref().join(filename);
+
+                    if url.scheme() == "file" {
+                        // Just copy the file.
+                        let src = match url.to_file_path() {
+                            Err(_) => Err(BuildError::NoUrlFile(url.clone()))?,
+                            Ok(s) => s,
+                        };
+
+                        // Eschew std::fs::copy for better error messages.
```

Comment:
There's a crate called [fs-err](https://docs.rs/fs-err/latest/fs_err/) that is essentially a wrapper over `std::fs` but with reasonable error messages! You might find it useful.

It'd be something like this:

```rust
use fs_err as fs;

// ..

fs::copy(src, dst)?;
```

Reply:
That's handy, but it sadly doesn't have a wrapper for [io::copy](https://doc.rust-lang.org/stable/std/io/fn.copy.html), only for [fs::copy](https://doc.rust-lang.org/stable/std/fs/fn.copy.html).

Reply:
Also I just tried it, and alas it is not a perfect drop-in replacement for `fs`. `fs::File::open`, for example, has this signature:

```rs
pub fn open<P: AsRef<Path>>(path: P) -> io::Result<File>
```

While fs-err has:

```rs
pub fn open<P>(path: P) -> Result<Self, io::Error>
    where
        P: Into<PathBuf>,
    {
```

I assume it mostly works, but not when the input is _defined_ as `AsRef<Path>`, as in the `unpack` method I just added in #7.